cgsmiles/read_cgsmiles.py

```diff
@@ -142,12 +146,22 @@ def read_cgsmiles(pattern):
 
         # here we check if the atom is followed by a cycle marker
         # in this case we have an open cycle and close it
-        if stop < len(pattern) and pattern[stop].isdigit() and pattern[stop] in cycle:
-            cycle_edge = (current, cycle[pattern[stop]])
-        # we open a cycle
-        elif stop < len(pattern) and pattern[stop].isdigit():
-            cycle_edge = None
-            cycle[pattern[stop]] = current
+        for token in pattern[stop:]:
+            # we close a cycle
+            if token.isdigit() and token in cycle:
+                cycle_edges.append((current, cycle[token]))
+            # we open a cycle
+            elif token.isdigit():
+                cycle[token] = current
+            # we close a cycle with the % syntax
+            elif token == "%" and _get_percent(pattern, stop) in cycle:
+                cycle_edges.append((current, cycle[_get_percent(pattern, stop)]))
+                break
+            elif token == "%":
+                cycle[_get_percent(pattern, stop)] = current
+                break
```

Comment:
```suggestion
            if token.isdigit() and token in cycle:
                cycle_edges.append((current, cycle[token]))
                del cycle[token]
            # we open a cycle
            elif token.isdigit():
                cycle[token] = current
            # we close a cycle with the % syntax
            elif token == "%" and _get_percent(pattern, stop) in cycle:
                # This would also be a good place for a walrus assignment
                cycle_edges.append((current, cycle[_get_percent(pattern, stop)]))
                del cycle[_get_percent(patter, stop)]
                break
            elif token == "%":
                cycle[_get_percent(pattern, stop)] = current
                break
```
In regular smiles you can reuse cycle numbers!

Also, can cycle edges have a bond order that is not 1? It is bad practice, but allowed by opensmiles

Reply:
@pckroon can you expand on the walrus?

Reply:
changed in new version and PR